backends/memcheck/ceed-memcheck-restriction.c

```diff
@@ -0,0 +1,801 @@
+// Copyright (c) 2017-2022, Lawrence Livermore National Security, LLC and other CEED contributors.
+// All Rights Reserved. See the top-level LICENSE and NOTICE files for details.
+//
+// SPDX-License-Identifier: BSD-2-Clause
+//
+// This file is part of CEED:  http://github.com/ceed
+
+#include <ceed.h>
+#include <ceed/backend.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "ceed-memcheck.h"
+
+//------------------------------------------------------------------------------
+// Set backend strides
+//------------------------------------------------------------------------------
+static inline int CeedElemRestrictionGetBackendStrides_Memcheck(CeedElemRestriction rstr, CeedInt l_vec_index, CeedInt strides[3]) {
+  CeedInt elem_size, num_comp, num_elem;
+
+  CeedCallBackend(CeedElemRestrictionGetElementSize(rstr, &elem_size));
+  CeedCallBackend(CeedElemRestrictionGetNumComponents(rstr, &num_comp));
+  CeedCallBackend(CeedElemRestrictionGetNumElements(rstr, &num_elem));
+  switch (l_vec_index % 4) {
+    case 0:  // Contiguous by component, then node
+      strides[0] = num_comp;
+      strides[1] = 1;
+      strides[2] = num_comp * elem_size;
+      break;
+    case 1:  // Contiguous by component, then element
+      strides[0] = num_comp * num_elem;
+      strides[1] = 1;
+      strides[2] = num_comp;
+      break;
+    case 2:  // Contiguous by element, then node
+      strides[0] = num_elem;
+      strides[1] = num_elem * elem_size;
+      strides[2] = 1;
+      break;
+    case 3:  // Contiguous by element, then component
+      strides[0] = num_elem * num_comp;
+      strides[1] = num_elem;
+      strides[2] = 1;
+      break;
+      /**
+      // CPU default
+      strides[0] = 1;
+      strides[1] = num_elem * elem_size;
+      strides[2] = elem_size;
+      // GPU default
+      strides[0] = 1;
+      strides[1] = elem_size;
+      strides[2] = elem_size * num_comp;
+     **/
+  }
```

Comment:
4 different orderings is probably overkill, but I was trying to avoid some accidental correspondence between vectors that masks a bug. This here should be the 6 possible dense orderings, so I avoided the CPU and GPU default layouts for l-vectors.